### DISCUSSION TITLE
[Immunization] Defining Immunization dosages and propagating them to all the relevant resources | Content Changes

### DISCUSSION THREAD

**Discussion by dubdabasoduba** [opensrp/fhircore#2765] [General]
# Context

- We need to record immunization dosages in a way that allows for proper querying and indicator calculation. Currently, we use the Task `description` field for filters when calculating indicators, but this is prone to breaking if the field's text changes.
- To address this, we will use the `Immunization.protocolApplied[x].doseNumber` field to store dosage information. This requires a few steps:
    - Adding dosage information to the Task resource.
    - Passing the dosage information from the Task to the StructureMap (SM) that generates the Immunization resource.
    - Adding the dosage to the Immunization resource.

## How to Add Dosage Information to the Task

- Obtain the dosage from the **Immunization PlanDefinition**. Here is an example of how to use this dosage in a StructureMap (SM):
`task.description = evaluate(definition, $this.product.text + ' ' + currentDose.select(sequence.toString()))`
- Update the SM that generates the tasks by adding an entry in the `Task.input` attribute. This input will hold the dose number. Here is an example:
  ```
  {
    "type": {
      "coding": [
        {
          "system": "http://snomed.info/sct",
          "code": "260911001",
          "display": "Dosage (attribute)"
        }
      ]
    },
    "value": "<dosage>"
  }
  ```
- Make sure to replace `<dosage>` with the actual dosage number.

### Through the Record Single Immunization Questionnaire

- Retrieve the `Task.input` created above from the patient profile using the following rule: `Task.input.where(type.coding.select(code) = '260911001').value`
- Pass this information to the Questionnaire through the `PREPOPULATE` functionality, as shown in [[this example](https://github.com/onaio/fhir-resources/blob/8b429e51f94f10078d294ccf1b6f74e4d3751ed8/ecbis_liberia/app_configurations/ecbis_cha_preview/profiles/default_profile_config.json#L1937-L1966)](https://github.com/onaio/fhir-resources/blob/8b429e51f94f10078d294ccf1b6f74e4d3751ed8/ecbis_liberia/app_configurations/ecbis_cha_preview/profiles/default_profile_config.json#L1937-L1966).
- The dosage information will be added to a hidden field in the Questionnaire, as demonstrated in [[this example](https://github.com/onaio/fhir-resources/blob/8b429e51f94f10078d294ccf1b6f74e4d3751ed8/ecbis_liberia/questionnaire/follow_up_visits/child_immunization_record_sinlge.json#L93-L104)](https://github.com/onaio/fhir-resources/blob/8b429e51f94f10078d294ccf1b6f74e4d3751ed8/ecbis_liberia/questionnaire/follow_up_visits/child_immunization_record_sinlge.json#L93-L104).
- Once the form is filled out, the dosage value will be available on the QuestionnaireResponse and can be used on the extraction SM.

### Through the Record All Questionnaire

- This questionnaire contains a Date question for each immunization at each dose.
- Update the questionnaire to add a dosage question. Each of the dosage questions should have the dosage as a default value.
- This field should not be editable or visible to the user.

## How to add dosage to the Immunization Resource

- Extract the dosage information from the respective QRs and use it to update the `immunization` resource after recording an immunization on the app.
- Add the dosage information to the `Immunization.protocolApplied[x].doseNumber` attribute. For example: `{"protocolApplied":[{"doseNumber":2}]}`

### Tasks resource
```json
{
  "resourceType": "Task",
  "id": "a9100c01-c84b-404f-9d24-9b830463a152",
  "identifier": [
    {
      "use": "official",
      "value": "a20e88b4-4beb-4b31-86cd-572e1445e5f3"
    }
  ],
  "basedOn": [
    {
      "reference": "CarePlan/28d7542c-ba08-4f16-b6a2-19e8b5d4c229"
    }
  ],
  "partOf": {
    "reference": "Task/650203d2-f327-4eb4-a9fd-741e0ce29c3f"
  },
  "status": "requested",
  "intent": "plan",
  "priority": "routine",
  "code": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "33879002",
        "display": "Administration of vaccine to produce active immunity (procedure)"
      }
    ]
  },
  "description": "OPV 1 at 6 wk vaccine",
  "for": {
    "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
  },
  "executionPeriod": {
    "start": "2021-11-12T00:00:00+00:00",
    "end": "2026-11-11T00:00:00+00:00"
  },
  "authoredOn": "2023-03-28T10:46:59+00:00",
  "requester": {
    "reference": "Practitioner/3812"
  },
  "owner": {
    "reference": "Practitioner/3812"
  },
  "reasonCode": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "111164008",
        "display": "Poliovirus vaccine"
      }
    ],
    "text": "OPV"
  },
  "reasonReference": {
    "reference": "Questionnaire/9b1aa23b-577c-4fb2-84e3-591e6facaf82"
  },
  "input": [
    {
      "type": {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "900000000000457003",
            "display": "Reference set attribute (foundation metadata concept)"
          }
        ]
      },
      "value": {
        "reference": "Task/650203d2-f327-4eb4-a9fd-741e0ce29c3f"
      }
    },
    {
      "type": {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "371154000",
            "display": "Dependent (qualifier value)"
          }
        ]
      },
      "value": 28
    },
    {
      "type": {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "260911001",
            "display": "Dosage (attribute)"
          }
        ]
      },
      "value": "1"
    }
  ],
  "output": [
    {
      "type": {
        "coding": [
          {
            "system": "http://snomed.info/sct",
            "code": "41000179103",
            "display": "Immunization record (record artifact)"
          }
        ]
      },
      "value": {
        "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
      }
    }
  ]
}
```

### Immunization resource
```json
{
  "resourceType": "Immunization",
  "id": "41921cfe-5074-4eec-925f-2bc581237660",
  "identifier": {
    "use": "official",
    "value": "6a637a79-df7b-4cc9-93b2-f73f965c31ab"
  },
  "status": "completed",
  "patient": {
    "reference": "Patient/3e3d698a-4edb-48f9-9330-2f1adc0635d1"
  },
  "encounter": {
    "reference": "Encounter/14e2ae52-32fc-4507-8736-1177cdaafe90"
  },
  "occurrenceDateTime": "2021-10-01T00:00:00+00:00",
  "protocolApplied": [
    {
      "doseNumber": 2
    }
  ]
}
```

**Comment by qiarie** [comment]
@dubdabasoduba @f-odhiambo In the description below, does `dosage` mean the same as sequence number (e.g. `0`) for the vaccine or does it include the vaccine name such as `OPV 0`

>Update the SM that generates the tasks by adding an entry in the Task.input attribute. This input will hold the dose number. Here is an example:
{"type":{"coding":[{"system":"<http://snomed.info/sct","code":"260911001","display":"Dosage> (attribute)"}]},"value":"<dosage>"}
Make sure to replace <dosage> with the actual dosage number.

**Reply by f-odhiambo** [reply]
IMO I think that would be a yes , as this is the same variable we use in the SM

**Comment by Lentumunai-Mark** [comment]
@dubdabasoduba @f-odhiambo 
On the record all questionnaire we have[ this ](https://github.com/onaio/fhir-resources/blob/8b429e51f94f10078d294ccf1b6f74e4d3751ed8/ecbis_liberia/questionnaire/follow_up_visits/child_immunization_record_all.json#L118) field vaccines code group which handles `Update the questionnaire to add a dosage question. Each of the dosage questions should have the dosage as a default value.`

Here is a sample JSON response

```json
{
      "linkId": "vaccines codes group",
      "text": "Vaccine codes",
      "item": [
        {
          "linkId": "oral polio 1",
          "answer": [
            {
              "valueCoding": {
                "system": "http://snomed.info/sct",
                "code": "111164008",
                "display": "Poliovirus vaccine"
              }
            }
          ]
        },
        {
          "linkId": "penta 1",
          "answer": [
            {
              "valueCoding": {
                "system": "http://snomed.info/sct",
                "code": "429511000124109",
                "display": "Tetanus + diphtheria + acellular pertussis vaccine"
              }
            }
          ]
        },
        {
          "linkId": "pneumo 1",
          "answer": [
            {
              "valueCoding": {
                "system": "http://snomed.info/sct",
                "code": "12866006",
                "display": "Pneumococcal vaccine"
              }
            }
          ]
        },
        {
          "linkId": "rota 1",
          "answer": [
            {
              "valueCoding": {
                "system": "http://snomed.info/sct",
                "code": "415354003",
                "display": "Rotavirus vaccine"
              }
            }
          ]
        }
      ]
    }
```
My question is if we can possibly utilize this on the record all?
